### PR TITLE
Add the httpsprovider component

### PR DESCRIPTION
**Pull request by rapphil**
**Description:** Add the httpsprovider component that allows the collector to load configuration from HTTP servers using the HTTPS protocol.

This PR is based on this work https://github.com/junhaoyu-aws/opentelemetry-collector/pull/4

A bit of refactoring was made and an internal configurablehttpprovider was created. This component is used by both the httpsprovider and the httpprovider. The only difference is the configuration of the transport layer.

**Testing:** Unit tests using httptest and self signed certificates were added.

**Documentation:** A readme was added for the httpsprovider component.

### CODE REVIEW

File: confmap/provider/httpprovider/README.md
Comment (bryan-aguilar) [inline confmap/provider/httpprovider/README.md:19]:
```suggestion
This config provider also has support to HTTPS. By default, It will rely on PKI in order to establish the connection between the client and the server. It also provides mechanisms for you to specify your own certificate authority as well as bypass hostname validation.
```

**Review by MovieStoreGuy — COMMENTED** [review]
Having read through the code a bit, I would prefer the optional settings to be configurable via flags like:

```
otelcol --tls-ca-cert-path=file://infra/testinfra:named=testenv --tls-ca-cert-path=file://infra/prodinfra:named=prodenv --config=https://internal.infra.io/config:tls-bundle=testenv
```

However, I am not sure how feasible that might be.

File: confmap/provider/httpprovider/provider.go
Comment (rapphil) [reply confmap/provider/httpprovider/provider.go]: today I learned. Thanks!

File: confmap/provider/httpprovider/provider.go
Comment (rapphil) [reply confmap/provider/httpprovider/provider.go:67]: Good points. I will do some research before answering.

Comment (rapphil) [reply confmap/provider/httpprovider/provider.go:67]: I took a look into the resolver code and it does not pass any parameter to the confmap providers, only the URI whose resource should be retrieved. 

I agree with your points and the first point has special impact on `OTEL_HTTP_PROVIDER_TLS_INSECURE_SKIP_VERIFY`, because you might want to skip name verification for only some of the domain names being fetched. 

For `OTEL_HTTP_PROVIDER_TLS_ADDITIONAL_CA_CERT` it is ok to operate as a singleton because that is a public certificate that will be added to a pool of certificate containing the certificates installed in the system. In fact, we the latter variable could be a comma separated list of CA certificates.

I don't have a solution or suggestion on how to remediate these shortcomings. I'm considering attending the SIG meeting to gather feedback.

File: confmap/provider/httpprovider/provider.go
Comment (Aneurysm9) [inline confmap/provider/httpprovider/provider.go:52]: I worry somewhat about having a non-niladic initializer.  None of the other providers (that I'm aware of) require arguments to their initializer and having arguments there may complicate any attempt to expand the collector builder to support specifying `confmap.Provider`s to be included in a build.

File: confmap/provider/httpprovider/provider.go
Comment (Aneurysm9) [reply confmap/provider/httpprovider/provider.go:67]: An alternative might be to use the URI fragment to provide configuration to the provider.  The fragment is not supposed to be sent with an HTTP request anyways, so it should be safe to use for client configuration.

```
https://example.com/myConfig#insecure=true&certPath=/path/to/cert
```

**Comment by mx-psi** [comment]
From the SIG, let's have a first PR with an https provider without any configurable parameters and move the discussion on how to configure a confmap provider to an issue.

File: confmap/provider/httpprovider/README.md
Comment (rapphil) [reply confmap/provider/httpprovider/README.md]: we decided to make this component not configurable in the first version, so this not applicable anymore.

File: confmap/provider/httpprovider/provider.go
Comment (rapphil) [reply confmap/provider/httpprovider/provider.go]: The latest version uses niladic initializers for the http and https config map providers.

File: confmap/provider/httpprovider/provider.go
Comment (rapphil) [reply confmap/provider/httpprovider/provider.go]: FYI, I created this proposal https://github.com/open-telemetry/opentelemetry-collector/issues/6892

File: confmap/provider/httpsprovider/README.md
Comment (mx-psi) [inline confmap/provider/httpsprovider/README.md:1]: Could you format  the section titles as Markdown titles? e.g. here

```suggestion
### What is this new component httpsprovider?
```

(and also below)

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (mx-psi) [inline confmap/provider/internal/configurablehttpprovider/provider.go:78]: Why not wrap the readfile error for context?

```suggestion
			return nil, fmt.Errorf("unable to read CA from %q URI: %w", fmp.caCertPath, err)
```

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (mx-psi) [inline confmap/provider/internal/configurablehttpprovider/provider.go:123]: The `with err` is usually implicitly understood on Go
```suggestion
		return nil, fmt.Errorf("unable to download the file via HTTP GET for uri %q: %w ", uri, err)
```

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (mx-psi) [inline confmap/provider/internal/configurablehttpprovider/provider.go:135]:
```suggestion
		return nil, fmt.Errorf("fail to read the response body from uri %q: %w", uri, err)
```

File: confmap/provider/httpsprovider/README.md
Comment (rapphil) [reply confmap/provider/httpsprovider/README.md]: Done!

File: confmap/provider/internal/configurablehttpprovider/provider_test.go
Comment (mx-psi) [inline confmap/provider/internal/configurablehttpprovider/provider_test.go:222]:
```suggestion
			tsURL, err := url.Parse(ts.URL)
			require.NoError(t, err)
```

**Review by mx-psi — COMMENTED** [review]
Only one very small comment otherwise LGTM

**Review by mx-psi — APPROVED** [review]

File: confmap/provider/httpprovider/provider.go
Comment (Aneurysm9) [inline confmap/provider/httpprovider/provider.go:26]:
```suggestion
// One example for HTTP URI is: http://localhost:3333/getConfig
```

File: confmap/provider/httpsprovider/provider.go
Comment (Aneurysm9) [inline confmap/provider/httpsprovider/provider.go:24]:
```suggestion
// This Provider supports "https" scheme. One example of an HTTPS URI is: https://localhost:3333/getConfig
```

**Review by Aneurysm9 — APPROVED** [review]
A couple minor doc comments.  Not a big fan of `configurablehttpprovider` as a package name, but it's internal and easy to change so not blocking here.

**Comment by rapphil** [comment]
@codeboten, @bogdandrutu and @dmitryax, is there anything that I can do to get this merged?

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (bogdandrutu) [inline confmap/provider/internal/configurablehttpprovider/provider.go:97]: This func can be merged with getHTTPClient and getHTTPSClient. I think will actually get cleaner than having 2 very small funcs and a large one.

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (bogdandrutu) [inline confmap/provider/internal/configurablehttpprovider/provider.go:114]: Can this be called in the `New`?

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (bogdandrutu) [inline confmap/provider/internal/configurablehttpprovider/provider.go:110]: s/transport/scheme?

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (bogdandrutu) [inline confmap/provider/internal/configurablehttpprovider/provider.go:133]: defer resp.Body.Close()?

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (rapphil) [reply confmap/provider/internal/configurablehttpprovider/provider.go:114]: The reason why it is not created in `New` is because the https client can return errors during its creation. `New` does not propagate errors.

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (rapphil) [reply confmap/provider/internal/configurablehttpprovider/provider.go:131]: It is already in line 123. Am I missing any convention to where this should be?

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (rapphil) [reply confmap/provider/internal/configurablehttpprovider/provider.go:95]: For reference of how it would look like:
```go
// Get the client based on the type of scheme that was selected.
func (fmp *provider) getClient() (*http.Client, error) {
	switch fmp.scheme {
	case HTTPScheme:
		return &http.Client{}, nil
	case HTTPSScheme:
		pool, err := x509.SystemCertPool()

		if err != nil {
			return nil, fmt.Errorf("unable to create a cert pool: %w", err)
		}
	
		if fmp.caCertPath != "" {
			cert, err := os.ReadFile(filepath.Clean(fmp.caCertPath))
	
			if err != nil {
				return nil, fmt.Errorf("unable to read CA from %q URI: %w", fmp.caCertPath, err)
			}
	
			if ok := pool.AppendCertsFromPEM(cert); !ok {
				return nil, fmt.Errorf("unable to add CA from uri: %s into the cert pool", fmp.caCertPath)
			}
		}
	
		return &http.Client{
			Transport: &http.Transport{
				TLSClientConfig: &tls.Config{
					InsecureSkipVerify: fmp.insecureSkipVerify,
					RootCAs:            pool,
				},
			},
		}, nil
	default:
		return nil, fmt.Errorf("invalid scheme type: %s", fmp.scheme)
	}
}
```

I prefer to be more verbose and having the 2 extra functions because we are encapsulating the logic of creation based on  schema type. This makes the code easier to understand and maintain IMHO: If we need to add more features in the https client creation, we will only change that function. we will not have the overhead of looking into the http one and the switch statement.

On a side note, I think I will rename the functions from `getXxxx` to `createXxxx` because we are effective creating something.

Comment (bogdandrutu) [reply confmap/provider/internal/configurablehttpprovider/provider.go]: I disagree with this, jumping from a 2lines func to a 2 lines func is not readable at all. Also the create funcs don't need to be on the struct but can be standalone funcs.

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (bogdandrutu) [inline confmap/provider/internal/configurablehttpprovider/provider.go:56]: Unnecessary func, in the tests you can cast to provider.

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (bogdandrutu) [inline confmap/provider/internal/configurablehttpprovider/provider.go:126]: don't use hardcoded values, use `http.StatusOK` (https://go.dev/src/net/http/status.go)

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (rapphil) [reply confmap/provider/internal/configurablehttpprovider/provider.go]: Is this a merge blocker? If yes, I will address.

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (rapphil) [reply confmap/provider/internal/configurablehttpprovider/provider.go:56]: sure, but then this would cost the type checking right?  Is that ok?

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (Aneurysm9) [reply confmap/provider/internal/configurablehttpprovider/provider.go]: We're discussing non-exported methods on a non-exported type in an internal package.  Can we land this PR and deal with the color of the bikeshed later?

Comment (bogdandrutu) [reply confmap/provider/internal/configurablehttpprovider/provider.go]: @Aneurysm9 usually when I create a PR and is about bikeshed, the maintainer is always right to get my work merged. So I am not sure why debating this :)

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (bogdandrutu) [reply confmap/provider/internal/configurablehttpprovider/provider.go:56]: We are doing that for tests not for non-test code. In tests if you want to test a specific `provider` instance you can manually create that and not call `New*`

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (Aneurysm9) [reply confmap/provider/internal/configurablehttpprovider/provider.go]: I don't know why we're debating this either.  It doesn't matter.  You and I both know it doesn't matter.  Why put more effort into something that doesn't matter?  This has multiple approvals as-is.  Please merge it and create issues for any nits you want to pick.  They will be addressed.

Comment (rapphil) [reply confmap/provider/internal/configurablehttpprovider/provider.go]: @bogdandrutu I put your suggestion in place in the last commit.

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (rapphil) [reply confmap/provider/internal/configurablehttpprovider/provider.go]: Done.

File: confmap/provider/internal/configurablehttpprovider/provider.go
Comment (bogdandrutu) [inline confmap/provider/internal/configurablehttpprovider/provider.go:53]: No need to initialize `caCertPath: "", insecureSkipVerify: false` with their zero values.

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**
